### PR TITLE
Components with add item (+) button clear cached values when item is added

### LINKED ISSUE #103: Components with add item (`+`) button should clear cached values when item is added
A number of components allow new items to be added to the database by clicking a `+` button.  For example the `CropSelector` component allows a new crop to be added and the `LocationSelector` component allows a new location to be added:

image

The values displayed by these components are cached by the `farmosUtil.js` library when they are fetched via the API.  So if a new item (e.g. crop) is added it will not appear in the dropdown when returning to the form.

To observe this:
- Visit the Tray Seeding form
- Click `+` on crop
- Create a new crop "AAA"
- Return to the Tray Seeding form
- The new crop "AAA" does not appear in the list of crops.
- Clear the browser history
- Return to the Tray Seeding form
- The new crop "AAA" now appears in the list of crops.

The fix:
- Have each component with a `+` clear the associated cached items when the `+` is clicked.
  - This will force the component to make an API request for the content the next time it is rendered, and thus the new items will be added to the list.

The components that currently need to be fixed are:
- `components/CropSelector`
- `components/LocationSelector`
- `components/TraySizeSelector`

- Others may have been added since the creation of this issue.  So, review all of the components in the `components` directory to determine if any additional components need to be updated.

Tests should be added to the `*.behavior.comp.cy.js` test file to confirm that clicking the `+` clears the cached items. It is not necessary for these tests to add new items, but they must confirm that the cache has been cleared. Cached items can be cleared using functions in the `library/farmosUtil/farmosUtil.js` library.

### PR DESCRIPTION
**Pull request by Shahir-47**
**Pull Request Description**

Fixes the following components:

- `components/CropSelector`
- `components/LocationSelector`
- `components/TraySizeSelector`
- `components/EquipmentSelector`

Additionally, adds or modifies the component tests to confirm that clicking the `+` clears the cached items. 

Closes #103
---

**Licensing Certification**

FarmData2 is a [Free Cultural Work](https://freedomdefined.org/Definition) and all accepted contributions are licensed as described in the LICENSE.md file. This requires that the contributor holds the rights to do so. By submitting this pull request **I certify that I satisfy the terms of the [Developer Certificate of Origin](https://developercertificate.org/)** for its contents.

### CODE REVIEW

File: components/SelectorBase/SelectorBase.content.comp.cy.js [hunk 1]
```diff
@@ -141,17 +142,17 @@ describe('Test the default SelectorBase content', () => {
...
 
     cy.get('@readySpy')
       .should('have.been.calledOnce')
       .then(() => {
         cy.get('[data-cy="selector-add-button"]').should('exist');
-        cy.get('[data-cy="selector-add-button"]')
-          .should('have.attr', 'href')
-          .then((href) => href)
-          .should('eq', 'add/option/url');
+        cy.get('[data-cy="selector-add-button"]').click();
```
Comment (Shahir-47) [inline components/SelectorBase/SelectorBase.content.comp.cy.js]: No .click() needed

File: components/SelectorBase/SelectorBase.content.comp.cy.js [hunk 2]
```diff
@@ -130,8 +130,9 @@ describe('Test the default SelectorBase content', () => {
       });
   });
 
-  it('Test addOptionUrl prop', () => {
+  it('Test includeAddButton prop', () => {
     const readySpy = cy.spy().as('readySpy');
+    const addClickedSpy = cy.spy().as('addClickedSpy');
```
Comment (Shahir-47) [inline components/SelectorBase/SelectorBase.content.comp.cy.js]: remove the addClickedSpy

File: components/CropSelector/CropSelector.permissions.comp.cy.js [hunk 3]
```diff
@@ -9,21 +9,24 @@ describe('Test the permission based CropSelector content', () => {
...
     cy.get('@readySpy')
       .should('have.been.calledOnce')
       .then(() => {
-        cy.get('[data-cy="selector-add-button"]')
-          .should('have.attr', 'href')
-          .then((href) => href)
-          .should('eq', '/admin/structure/taxonomy/manage/plant_type/add');
+        cy.get('[data-cy="selector-add-button"]').should('exist');
+
+        // Simulate a click on the add button and check if the event is emitted
```
Comment (Shahir-47) [inline components/CropSelector/CropSelector.permissions.comp.cy.js]: get rid of the comment

File: components/CropSelector/CropSelector.permissions.comp.cy.js [hunk 4]
```diff
@@ -9,21 +9,24 @@ describe('Test the permission based CropSelector content', () => {
 
   it('Admin can add crops and url is correct', () => {
     const readySpy = cy.spy().as('readySpy');
+    const addClickedSpy = cy.spy().as('addClickedSpy');
```
Comment (Shahir-47) [inline components/CropSelector/CropSelector.permissions.comp.cy.js]: get rid of the spy

File: components/CropSelector/CropSelector.permissions.comp.cy.js [hunk 5]
```diff
@@ -9,21 +9,24 @@ describe('Test the permission based CropSelector content', () => {
 
   it('Admin can add crops and url is correct', () => {
```
Comment (Shahir-47) [inline components/CropSelector/CropSelector.permissions.comp.cy.js]: change description

File: components/LocationSelector/LocationSelector.vue [hunk 6]
```diff
@@ -12,6 +12,8 @@
       v-bind:showValidityStyling="showValidityStyling"
```
Comment (Shahir-47) [inline components/LocationSelector/LocationSelector.vue:11]: Remove addLocationURL

File: components/LocationSelector/LocationSelector.vue [hunk 7]
```diff
@@ -314,6 +303,32 @@ export default {
...
+      if (
+        this.includeFields &&
+        (this.includeGreenhouses || this.includeGreenhousesWithBeds) &&
+        this.canCreateLand &&
+        this.canCreateStructure
+      ) {
+        farmosUtil.clearCachedFields();
+        farmosUtil.clearCachedGreenhouses();
+        farmosUtil.clearCachedBeds();
+
```
Comment (Shahir-47) [inline components/LocationSelector/LocationSelector.vue]: remove blank line

File: components/LocationSelector/LocationSelector.permissions.comp.cy.js [hunk 8]
```diff
@@ -21,11 +21,6 @@ describe('Test permissions based content in the LocationSelector', () => {
       .should('have.been.calledOnce')
       .then(() => {
         cy.get('[data-cy="location-selector"]').should('exist');
```
Comment (Shahir-47) [inline components/LocationSelector/LocationSelector.permissions.comp.cy.js]: check if the button exists

File: components/LocationSelector/LocationSelector.permissions.comp.cy.js [hunk 9]
```diff
@@ -21,11 +21,6 @@ describe('Test permissions based content in the LocationSelector', () => {
       .should('have.been.calledOnce')
```
Comment (Shahir-47) [inline components/LocationSelector/LocationSelector.permissions.comp.cy.js:21]: change the it description

File: components/LocationSelector/LocationSelector.addGreenhousesWithBedsContent.comp.cy.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
+import LocationSelector from '@comps/LocationSelector/LocationSelector.vue';
+
+describe('Test the default LocationSelector content', () => {
```
Comment (Shahir-47) [inline components/LocationSelector/LocationSelector.addGreenhousesWithBedsContent.comp.cy.js:3]: change the description

File: components/LocationSelector/LocationSelector.addGreenhousesContent.comp.cy.js [hunk 11]
```diff
@@ -0,0 +1,38 @@
+import LocationSelector from '@comps/LocationSelector/LocationSelector.vue';
+
+describe('Test the default LocationSelector content', () => {
```
Comment (Shahir-47) [inline components/LocationSelector/LocationSelector.addGreenhousesContent.comp.cy.js]: update description

File: components/LocationSelector/LocationSelector.addGreenhousesBehavior.comp.cy.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
+import LocationSelector from '@comps/LocationSelector/LocationSelector.vue';
+import * as farmosUtil from '@libs/farmosUtil/farmosUtil.js';
+
+describe('Test the LocationSelector component behavior', () => {
```
Comment (Shahir-47) [inline components/LocationSelector/LocationSelector.addGreenhousesBehavior.comp.cy.js]: change the description

File: components/LocationSelector/LocationSelector.addGreenhousesBehavior.comp.cy.js [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    cy.restoreLocalStorage();
+    cy.restoreSessionStorage();
+  });
+
+  afterEach(() => {
+    cy.saveLocalStorage();
+    cy.saveSessionStorage();
+  });
+
+  it('Clicking add greenhouses button goes to the add location form and clears the greenhouses cache', () => {
```
Comment (Shahir-47) [inline components/LocationSelector/LocationSelector.addGreenhousesBehavior.comp.cy.js]: change it comment

**Comment by Shahir-47** [comment]
Remove all the content tests for LocationSelector

File: components/LocationSelector/LocationSelector.addFieldsContent.comp.cy.js [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+      },
+    });
+
+    cy.get('@readySpy')
+      .should('have.been.calledOnce')
+      .then(() => {
+        cy.get('[data-cy="location-selector"]').should('exist');
+        cy.get('[data-cy="selector-add-button"]').should('exist');
+        cy.get('[data-cy="selector-add-button"]').click();
+        cy.wait('@urlIntercept').its('response.statusCode').should('eq', 200);
```
Comment (Shahir-47) [inline components/LocationSelector/LocationSelector.addFieldsContent.comp.cy.js]: take out the click and check for buttons existence. Put these tests back into the .content file

File: components/TraySizeSelector/TraySizeSelector.adminPermissions.comp.cy.js [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+      props: {
+        onReady: readySpy,
+      },
+    });
+
+    cy.get('@readySpy')
+      .should('have.been.calledOnce')
+      .then(() => {
+        cy.get('[data-cy="selector-add-button"]').should('exist');
+        cy.get('[data-cy="selector-add-button"]').click();
```
Comment (Shahir-47) [inline components/TraySizeSelector/TraySizeSelector.adminPermissions.comp.cy.js]: don't click. put them back into the same permission file

**Comment by braughtg** [comment]
When loading the Direct Seeding or the Transplanting pages I am seeing the following error reported in the console:

```
Uncaught (in promise) Error: Permission [object Object] does not exist.
    po http://farmos/sites/all/modules/farmdata2/uiUtil/uiUtil.js:2
    Tn http://farmos/sites/all/modules/farmdata2/_plugin-vue_export-helper/_plugin-vue_export-helper.js:13
    b http://farmos/sites/all/modules/farmdata2/_plugin-vue_export-helper/_plugin-vue_export-helper.js:13
    run http://farmos/sites/all/modules/farmdata2/_plugin-vue_export-helper/_plugin-vue_export-helper.js:9
    update http://farmos/sites/all/modules/farmdata2/_plugin-vue_export-helper/_plugin-vue_export-helper.js:13
    se http://farmos/sites/all/modules/farmdata2/_plugin-vue_export-helper/_plugin-vue_export-helper.js:13
    he http://farmos/sites/all/modules/farmdata2/_plugin-vue_export-helper/_plugin-vue_export-helper.js:13
    X http://farmos/sites/all/modules/farmdata2/_plugin-vue_export-helper/_plugin-vue_export-helper.js:13
    R http://farmos/sites/all/modules/farmdata2/_plugin-vue_export-helper/_plugin-vue_export-helper.js:13
uiUtil.js:2:65868
```

The cause of this error should be identified and resolved.

Note that the error does not occur when loading the Tray Seeding page, so comparison to the parts that deal with permissions may be one way to resolve this issue.  One immediate thought is that it may be due to some component that appears on the the Direct Seeding and Transplanting forms that does not appear on the Direct Seeding page.  So possibly the Equipment Selector?

File: components/SelectorBase/SelectorBase.content.comp.cy.js [hunk 16]
```diff
@@ -151,8 +149,7 @@ describe('Test the default SelectorBase content', () => {
       .should('have.been.calledOnce')
       .then(() => {
         cy.get('[data-cy="selector-add-button"]').should('exist');
-        cy.get('[data-cy="selector-add-button"]').click();
-        cy.get('@addClickedSpy').should('have.been.calledOnce');
+        cy.get('[data-cy="selector-add-button"]');
```
Comment (braughtg) [inline components/SelectorBase/SelectorBase.content.comp.cy.js]: Remove this line, which does not make an assertion.

**Review by braughtg — CHANGES_REQUESTED** [review]

**Comment by braughtg** [comment]
> When loading the Direct Seeding or the Transplanting pages I am seeing the following error reported in the console:

This seems to have been a transitory issue that has resolved.

**Pull request merged**